### PR TITLE
Add output_writer_iterator

### PR DESCRIPTION
**Pull request by karthikeyann**
This iterator serves as a generalized version of output iterator.
Typical usage is
`output_writer_iterator(index_iterator, [result_begin](auto i, auto v) { result_begin[i]=v;});`

Sometimes output iterator has to be complex. This output writer iterator can help achieve writing to complex data structures using binary function object or lambda which accepts an index and an value to write.
Example given in documentation is achieving functionality similar to `std::bitset` using a `thrust::device_vector`.

### CODE REVIEW

File: thrust/iterator/output_writer_iterator.cuh [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Output writer iterator
+#pragma once
+
+#include <thrust/iterator/iterator_adaptor.h>
+
+namespace thrust {
```
Comment (allisonvacanti) [inline thrust/iterator/output_writer_iterator.cuh:22]: Use the `THRUST_NAMESPACE_BEGIN` / `THRUST_NAMESPACE_END` macros instead.

```suggestion
THRUST_NAMESPACE_BEGIN
```

File: thrust/iterator/output_writer_iterator.cuh [hunk 2]
```diff
@@ -0,0 +1,140 @@
+/*
```
Comment (allisonvacanti) [inline thrust/iterator/output_writer_iterator.cuh:1]: This file will need to be renamed. Thrust uses the `.h` extension instead of `.cuh` for all headers.

File: thrust/iterator/output_writer_iterator.cuh [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+ *   __device__ void operator()(size_t i, bool x)
+ *   {
+ *     if (x)
+ *       set_bit(i);
+ *     else
+ *       clear_bit(i);
+ *   }
+ * };
+ *
+ * thrust::device_vector<int> v(1, 0x0000);
```
Comment (allisonvacanti) [inline thrust/iterator/output_writer_iterator.cuh:83]: Nitpick: It seems odd to initialize a 32-bit value with a 16-bit hex 0. For clarity, consider:

```suggestion
 * thrust::device_vector<int> v(1, 0x00000000);
```

(or just `thrust::device_vector<int> v(1, 0);`)

File: thrust/iterator/output_writer_iterator.cuh [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+ * auto result_begin = thrust::make_output_writer_iterator(thrust::make_counting_iterator(0),
+ *                                                 set_bits_field{v.data().get()});
+ * auto value = thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+ *   [] __device__ (int x) {   return x%2; });
+ * thrust::copy(thrust::device, value, value+32, result_begin);
+ * int(v[0]); // returns 0xaaaaaaaa;
```
Comment (allisonvacanti) [inline thrust/iterator/output_writer_iterator.cuh:89]: No need to cast here, `v[0]` is already an `int`.

```suggestion
 * assert(v[0] == 0xaaaaaaaa);
```

File: thrust/iterator/output_writer_iterator.cuh [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  }
+};
+
+template <typename BinaryFunction, typename Iterator>
+output_writer_iterator<BinaryFunction, Iterator> __host__ __device__
+make_output_writer_iterator(Iterator out, BinaryFunction fun)
+{
+  return output_writer_iterator<BinaryFunction, Iterator>(out, fun);
+}  // end make_output_writer_iterator
+}  // namespace thrust
```
Comment (allisonvacanti) [inline thrust/iterator/output_writer_iterator.cuh:140]: Need to use the namespace macro here:

```suggestion
THRUST_NAMESPACE_END
```

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
In addition to the inline comments, I have some other suggestions for this PR:

------

Rather than having an output-only iterator, let's update this to support both input and output by replacing `BinaryFunction` with `InputBinaryFunction` and `OutputBinaryFunction`, where the two functions are inverse operations. This will avoid the current mess we have with the transform iterators (we currently provide `transform_iterator`, `transform_output_iterator` and `transform_input_output_iterator` for legacy reasons).

------

The name `writer_iterator` isn't very clear IMO, and won't make much sense once this supports input and output. Some alternative ideas:

- `proxy_iterator`
- `functor_iterator`
- `invoke_iterator`
- `arg_iterator`

I'm not sure I like these any better, though. Let's think about this a bit more, and maybe see if anyone else has some ideas.

Edit: `decorate_iterator` or something similar (`decorating_`, `decorated`_, etc) might be a good fit, since this effectively implements the "decorator" design pattern over an existing iterator. I've started a discussion on this with the rest of the team to see if we can find something better.

------

This will need tests and examples covering the input/output/both usecases.

------

So to summarize:

- [ ] Change to input/output iterator.
- [ ] Let's work out a more meaningful name for this.
- [ ] Add tests and examples of using this iterator for input, output, and both.

**Comment by allisonvacanti** [comment]
Updating with some internal conversations:

-----

@harrism suggested `indexed_iterator`. This seems like the best candidate so far IMO, since the key feature of this iterator is that it provides the current index to the functor(s).

------

> What would the input and input/output cases look like?

Input:

```cpp
struct adj_diff_functor
{
  const int *input;
​
  __device__
  int operator()(std::size_t i) const
  {
    return input[i + 1] - input[i];
  }
};
​
thrust::device_vector<int> data{0, 1, 4, 7, 9, 12};
auto iter = thrust::make_indexed_input_iterator(adj_diff_functor{data.data()});
auto iter_end = iter + data.size() - 1;
​
// [iter, iter_end) -> {1, 3, 3, 2, 3}
```

@benbarsdell correctly points out that this could be implemented using existing counting + transform iterators. 

Input/output is a more interesting case that cannot be implemented using other existing fancy iterators:

```cpp
struct input_functor
{
  // Produces ones[i] + 10*tens[i]
  const int *ones; // Always 0-9
  const int *tens;
  
  __device__
  int operator()(std::size_t i) const
  {
    return ones[i] + 10 * tens[i];
  }
};

struct output_functor
{
  // Decomposes input values to perform the inverse of input_functor
  const int *ones; // Always 0-9
  const int *tens;
  
  __device__
  void operator()(std::size_t i, int val) const
  {
    ones[i] = val % 10;
    tens[i] = val / 10;
  }
};

thrust::device_vector<int> ones = {1, 5, 3, 2, 6};
thrust::device_vector<int> tens = {3, 2, 6, 7, 3};

assert(ones.size() == tens.size())
auto iter = thrust::make_indexed_iterator(input_functor{ones.data(), tens.data()},
                                          output_functor{ones.data(), tens.data()});
auto iter_end = iter + ones.size();

// [iter, iter_end) -> {31, 25, 63, 72, 36}

iter[0] = 19;
iter[1] = 42;
iter[2] = 98;
iter[3] = 132;
iter[4] = 15;

// ones -> {9, 2, 8, 2, 5}
// tens -> {1, 4, 9, 13, 1}
```

**Comment by harrism** [comment]
> > @harrism suggested `indexed_iterator`. This seems like the best candidate so far IMO, since the key feature of this iterator is that it provides the current index to the functor(s).
> 
> Looking to [`thrust::tabulate`](https://thrust.github.io/doc/group__transformations_ga0ddf108cc4d1109addeed3251824f951.html) for inspiration, how about `tabulate_iterator`?

I like `tabulate_iterator` for the case where you cannot provide an iterator that generates the indices.  But if, as in @karthikeyann 's original example, we enable passing an iterator that generates the indices, these may not just be sequential/counting. In that case `indexed_iterator` may be a better name. For example you may want to pass a counting transform iterator to generate indices with a stride, or a reverse counting iterator, etc.

**Comment by jrhemstad** [comment]
> > Decorator pattern allows a user to add new functionality to an existing object without altering its structure.
> > This pattern creates a decorator class which wraps the original class and provides additional functionality keeping class methods signature intact.
> 
> This is an iterator that wraps all accesses to an existing iterator with decorator functions.

What gives me pause about this is that this pattern/definition applies just as well to `transform_iterator`. 

In fact, the input iterator case for the `decorated_iterator` is exactly equivalent to a `transform_iterator`. The only difference is in assignment (or using it as an output iterator). 

I'm not sure the name `decorated_iterator` vs. `transform_iterator` conveys that difference. 

Now I'm thinking that maybe this new iterator can just be a new overload of `transform_iterator` as opposed to a wholly new iterator.

**Comment by allisonvacanti** [comment]
> What gives me pause about this is that this pattern/definition applies just as well to `transform_iterator`.

It is a pretty generic name, and also very similar to the existing `iterator_adaptor`. I'd love to find something more specific.

> Now I'm thinking that maybe this new iterator can just be a new overload of transform_iterator as opposed to a wholly new iterator.

Hmm, perhaps. Let's compare their user interfaces directly:

```cpp
{ // Transform:
  T = value_t<Iter>;
  U InFunc(T); // Used for reads
  T OutFunc(U); // Used for writes
  auto thrust::make_transform_iterator(Iter, InFunc, OutFunc);
}

{ // New Iterator
  T = value_t<Iter>;
  U InFunc(T); // Used for reads
  void OutFunc(T, U); // Used for writes
  auto thrust::make_XXX_iterator(Iter, InFunc, OutFunc);
}
```

The key difference is in `OutFunc`. While `transform` will derive a new value from the assignment arg and write it to `Iter`, the new iterator will never modify `Iter`'s data. `Iter`'s data is only used as an input to modify state captured in the functors.

**Comment by allisonvacanti** [comment]
> Iter's data is only used as an input to modify state captured in the functors.

Which implies that this iterator is only useful with some external state, which could potentially be abstracted out:

```cpp
{ // New Iterator with abstracted state:
  using T = value_t<Iter>;
  U InFunc(State, T); // Used for reads
  void OutFunc(State, T, U); // Used for writes
  auto thrust::make_XXX_iterator(Iter, State, InFunc, OutFunc);
}
```

This *could* be used to simplify the functors so they don't have to capture the same state in their members:

```cpp
{ // current
  struct input_functor
  {
    State *state;  
    U operator()(T) const;
  };

  struct output_functor
  {
    State *state;  
    void operator()(T, U) const;
  };

  State *state = ...;
  auto iter = thrust::make_XXX_iterator(
    indicies,
    input_functor{state},
    output_functor{state});
}

{ // with state abstraction
  struct input_functor
  {
    U operator()(State, T) const;
  };

  struct output_functor
  {
    void operator()(State, T, U) const;
  };

  State *state = ...;
  auto iter = thrust::make_XXX_iterator(
    indicies,
    state,
    input_functor{},
    output_functor{});
}
```

I'm unsure whether this abstraction would be worthwhile, but I think a key point here is that this iterator does require extra state to do anything interesting.

**Comment by jrhemstad** [comment]
> I think a key point here is that this iterator does require extra state to do anything interesting.

Yeah, I noticed that too. I'm not a fan of the fact that the binary callable doesn't return anything, which means it has to have side-effects to do anything useful.

**Comment by allisonvacanti** [comment]
> I'm not a fan of the fact that the binary callable doesn't return anything, which means it has to have side-effects to do anything useful.

I don't think that'll be a problem. For instance, `std::for_each` uses a functor with no return value and relies on side effects, so it's not uncommon.

Do you see a different way to implement this functionality?

**Comment by karthikeyann** [comment]
The main reason for creating this iterator is to provide a generalized output iterator. 
For input iterator, we can create `counting_iterator` and use a `make_transform_iterator` to translate it to access any datastructure. 
But for creating output writer, AFAIK there is no way right now, except for the data structure to provide itself.

`BinaryFunction` is present only because first argument is supposed to the index. Instead of providing an pre-defined iterator index, I made it as an argument (which again gives more flexibility). Or else it could be `UnaryFunction`. If it is unary, `tabulate_iterator` or `indexed_iterator` name would make sense. but again it's missing _output_ in the name.

Also, mixing input and output together would make it more than what it's designed for. My vote would be to limit this iterator's functionality to behave like output iterator only, and add new combined input and output iterator seperately.

**Comment by jrhemstad** [comment]
> The main reason for creating this iterator is to provide a generalized output iterator.
> For input iterator, we can create counting_iterator and use a make_transform_iterator to translate it to access any datastructure.
> But for creating output writer, AFAIK there is no way right now, except for the data structure to provide itself.

@karthikeyann's description reminded me that I explored an alternative for this approach awhile back in https://github.com/NVIDIA/thrust/issues/1588#issuecomment-1012435530. 

However, coming back to this after some time, it occurs to me that we are effectively just trying to indirectly recreate [projections](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4128.html#algorithms-should-take-invokable-projections) via a complicated iterator. 

That tells me there is probably a far more natural way to express this. 

Curious for @ericniebler 's thoughts.

**Comment by karthikeyann** [comment]
@jrhemstad In https://github.com/NVIDIA/thrust/issues/1588#issuecomment-1012435530 godbolt example, it still uses `f.begin()` as starting point for the output iterator. This output iterator is transformed using projections. Instead of `f.begin()`, can you use a `counting_iterator(0)`?
